src/wasm/wasm-type.cpp

```diff
@@ -2203,6 +2313,239 @@ bool FiniteShapeEquator::eq(const Rtt& a, const Rtt& b) {
   return a.depth == b.depth && eq(a.heapType, b.heapType);
 }
 
+size_t RecGroupHasher::operator()() const {
+  size_t digest = wasm::hash(group.size());
+  for (auto type : group) {
+    hash_combine(digest, topLevelHash(type));
+  }
+  return digest;
+}
+
+size_t RecGroupHasher::topLevelHash(HeapType type) const {
+  size_t digest = wasm::hash(type.isBasic());
+  if (type.isBasic()) {
+    wasm::rehash(digest, type.getID());
+  } else {
+    hash_combine(digest, hash(*getHeapTypeInfo(type)));
+  }
+  return digest;
+}
+
+size_t RecGroupHasher::innerHash(HeapType type) const {
+  // Do not recurse into the structure of this child type, but rather hash it as
+  // an index into a rec group. Only take the rec group identity into account if
+  // the child is not a member of the top-level group because in that case the
+  // group may not be canonicalized yet.
+  size_t digest = wasm::hash(type.getRecGroupIndex());
+  auto currGroup = type.getRecGroup();
+  if (currGroup != group) {
+    wasm::rehash(digest, currGroup.getID());
+  }
```

Comment:
I think it would be good to add an else here with a rehash of something that cannot overlap with the getID in the if. Without that, if this hash is combined with other things then it can compose in ways that happen to overlap.

Reply:
I don't believe there's any way such a composition could exist (modulo weaknesses in the hash function that could cause collisions in any other circumstance). In other words, not having a rehash here is just as good as adding a rehash of some arbitrary constant I could come up with.

Reply:
If you have a struct with 3 child hash types, then we'll hash them in sequence. Without ensuring the same "structure" of hashing operations, that might end up overlapping with a struct with say 4 children, if the one with 3 does an extra rehashing for one of its children.

Is `getID()` a pointer? Then rehashing with a 0 would be a value that cannot collide.

Here are places where we do similar things:

https://github.com/WebAssembly/binaryen/blob/f2d53918155cfe6e7fc333be22f25ae3776b1438/src/ir/ExpressionAnalyzer.cpp#L275-L277

https://github.com/WebAssembly/binaryen/blob/f2d53918155cfe6e7fc333be22f25ae3776b1438/src/ir/ExpressionAnalyzer.cpp#L343-L346

Or are you saying there is a logical reason stemming from the isorecursive structure that prevents the problem here? I may be just missing that then...

Reply:
In the locations you linked to, it is possible that the code path is reached _n_ times in a row and without hashing in a constant, different values of _n_ would result in the same hash. In this case, the hashing of the index above this branch ensures that calling this method _n_ times returns different results for different values of _n_. Given that, not modifying the hash is just as good as hashing in a constant.

Reply:
What you say is right, but it is an additional issue to the one I mentioned. Maintaining the "structural form" of hashing has been noticed to help in practice. I mean, we don't want these to hash the same:
```
[1 2 3 4]
[1 2 3] [4]
```
The second line has a shorter logical first item and a second short item that happens to contain 4. The same 4 numbers are hashed in each case, but logically the result should differ. This is solved by ensuring the structure to keep the right size:
```
[1 2 3 4]
[1 2 3 0] [4]
```
Here `0` is a value we know cannot overlap with other things possible in that position.